scripts/codegen/modules/devicedeclare.py

```diff
@@ -0,0 +1,391 @@
+    mapping = {}
+    # add preset mapping
+    mapping.update(preset)
+    # add device properties from device tree
+    mapping.update(codegen.edts().device_properties_flattened(device_id))
+    # add specific device declaration vars/ constants
+    mapping['device-name'] = mapping.get('device-name',
+                                         _device_name(device_id))
+    mapping['driver-name'] = mapping.get('driver-name',
+        codegen.edts().device_property(device_id, 'label').strip('"'))
+    mapping['device-data'] = mapping.get('device-data',
+        "{}_data".format(mapping['device-name']).lower())
+    mapping['device-config-info'] = mapping.get('device-config-info',
+        "{}_config".format(mapping['device-name']).lower())
+    mapping['device-config-irq'] = mapping.get('device-config-irq',
+        "{}_config_irq".format(mapping['device-name']).lower())
+    # config vars
+    mapping.update(codegen.config_properties())
+    substituted = _DeviceLocalTemplate(template).safe_substitute(mapping)
+
+    # substitute device global placeholders ${<device-id>:<property_path>}
+    #
+    # we need a second substitude to allow for device indirections
+    # ${${<property_path for device id>}:<property_path>}
+    mapping = {}
+    for device_id in codegen.edts()['devices']:
+        path_prefix = device_id + ':'
+        mapping.update(codegen.edts().device_properties_flattened(device_id,
+                                                                  path_prefix))
+        # add specific device declaration vars/ constants
+        try:
+            mapping[path_prefix + 'device-name'] = _device_name(device_id)
+            mapping[path_prefix + 'driver-name'] = \
+                codegen.edts().device_property(device_id, 'label').strip('"')
+        except:
+            # will be obvious if any of this is needed, just skip here
+            pass
+    mapping.update(codegen.config_properties())
+
+    # add aliases to mapping
+    aliases_mapping = {}
+    for property_path, property_value in mapping.items():
+        for alias_property_path, alias in _property_path_aliases:
+            if property_path.endswith(alias_property_path):
+                property_path = property_path[:-len(alias_property_path)] \
+                                + alias
+                aliases_mapping[property_path] = property_value
+    mapping.update(aliases_mapping)
+
+    substituted = _DeviceGlobalTemplate(substituted).safe_substitute(mapping)
+
+    return substituted
+
+
+#
+# @return True if device is declared, False otherwise
+def device_declare_single(device_config,
+                          driver_name,
+                          device_init,
+                          device_level,
+                          device_prio,
+                          device_api,
+                          device_info):
+    device_configured = codegen.config_property(device_config, '<not-set>')
```

Comment:
I had that in the beginning but I ran into the following problem:

If config is not set it may be that this is because the device instance is not part of the SoC variant (but supported by the driver for another variant). You can not find the device_id of a not existing device instance. The python code throws an exception.  

You may argument then - don´t throw the exception but silently ignore this. By doing so we get the same effect as the current code but it is less visible and you can not detect the critical error situation where there is a config but no active device. 

Thats why I decided to not use the #ifdef construct.

Reply:
> If config is not set it may be that this is because the device instance is not part of the SoC variant (but supported by the driver for another variant). You can not find the device_id of a not existing device instance.

We should not face definition inconsistencies at dts level. If we do, we've messed up somewhere.
It's up to device tree to hold this information, not Kconfig. DTS and (dts related generation), should define and generate what is coherent vs the SoC variant, then Kconfig should provide way to the user to control which subsystem to enable and which instance to compile.
At dts step, we should behave as if all Kconfig flags were on (and actually dts should define what are the possible instances Kconfig flags).